File: memq-client/src/main/java/com/pinterest/memq/client/commons/MemqLogMessageIterator.java
/**
 * Copyright 2022 Pinterest, Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.pinterest.memq.client.commons;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;
import java.util.logging.Logger;

import com.codahale.metrics.MetricRegistry;
import com.google.gson.JsonObject;
import com.pinterest.memq.client.commons.audit.Auditor;
import com.pinterest.memq.commons.BatchHeader;
import com.pinterest.memq.commons.CloseableIterator;
import com.pinterest.memq.commons.MemqLogMessage;
import com.pinterest.memq.commons.MessageId;
import com.pinterest.memq.core.utils.MemqUtils;

public class MemqLogMessageIterator<K, V> implements CloseableIterator<MemqLogMessage<K, V>> {

  private static final Logger logger = Logger
      .getLogger(MemqLogMessageIterator.class.getCanonicalName());
  public static final String METRICS_PREFIX = "memqConsumer";
  public static final String MESSAGES_PROCESSED_COUNTER_KEY = METRICS_PREFIX
      + ".messagesProcessedCounter";
  public static final String BYTES_PROCESSED_METER_KEY = METRICS_PREFIX + ".bytesProcessedCounter";
  protected DataInputStream uncompressedBatchInputStream;
  protected Deserializer<V> valueDeserializer;
  protected Deserializer<K> headerDeserializer;
  protected int messagesToRead;
  protected MetricRegistry metricRegistry;
  protected JsonObject currNotificationObj;
  protected int currentMessageOffset;
  protected DataInputStream stream;
  private int notificationPartitionId;
  private long notificationPartitionOffset;
  private long notificationReadTimestamp;
  private int objectSize;
  private Auditor auditor;
  private String cluster;
  private String topic;
  private MemqMessageHeader header;
  private byte[] messageIdHash;
  private int auditedMessageCount;
  private BatchHeader batchHeader = null;
  private String clientId;
  
  protected MemqLogMessageIterator() {
  }

  public MemqLogMessageIterator(String cluster,
                                String clientId,
                                DataInputStream stream,
                                JsonObject currNotificationObj,
                                Deserializer<K> headerDeserializer,
                                Deserializer<V> valueDeserializer,
                                MetricRegistry metricRegistry,
                                boolean skipHeaderRead,
                                Auditor auditor) throws IOException {
    this.cluster = cluster;
    this.clientId = clientId;
    this.stream = stream;
    this.currNotificationObj = currNotificationObj;
    this.auditor = auditor;
    try {
      this.topic = currNotificationObj.get(MemqLogMessage.INTERNAL_FIELD_TOPIC).getAsString();
      this.objectSize = currNotificationObj.get(MemqLogMessage.INTERNAL_FIELD_OBJECT_SIZE).getAsInt();
      this.notificationPartitionId = currNotificationObj
          .get(MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_PARTITION_ID).getAsInt();
      this.notificationPartitionOffset = currNotificationObj
          .get(MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_PARTITION_OFFSET).getAsLong();
      this.notificationReadTimestamp = currNotificationObj
          .get(MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_READ_TIMESTAMP).getAsLong();
    } catch (Exception e) {
    }
    this.headerDeserializer = headerDeserializer;
    this.valueDeserializer = valueDeserializer;
    this.metricRegistry = metricRegistry;
    if (!skipHeaderRead) {
      batchHeader = new BatchHeader(stream);
    }
    readHeaderAndLoadBatch();
  }

  @Override
  public boolean hasNext() {
    // no more notifications
    try {
      if (stream.available() <= 0) {
        // last batch remaining to read
        return messagesToRead > 0;
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    // more notifications or object not fully read
    return true;
  }

  @Override
  public MemqLogMessage<K, V> next() {
    if (messagesToRead > 0) {
      // still more MemqLogMessages in this batch
      try {
        return getMemqLogMessage();
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    } else {
      // no more MemqLogMessages in this batch
      try {
        if (stream.available() > 0) {
          // there are more batches, process new batch
          if (readHeaderAndLoadBatch()) {
            return next();
          }
        }
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    }
    throw new RuntimeException(new DataCorruptionException("No next"));
  }

  /**
   * Reads the next batch's header and returns whether there are bytes to read
   *
   * @return true if there are batch bytes to read, false otherwise
   */
  protected boolean readHeaderAndLoadBatch() {
    try {
      // since we are about to read a brand new batch therefore we should reset the
      // messageIdHash
      messageIdHash = null;
      auditedMessageCount = 0;
      header = readHeader();
      logger.fine(() -> "Message header:" + header);
      byte[] batch = new byte[header.getMessageLength()];
      stream.read(batch);
      if (!CommonUtils.crcChecksumMatches(batch, header.getCrc())) {
        // CRC checksum mismatch
        throw new RuntimeException(new DataCorruptionException("CRC checksum mismatch"));
      }
      if (uncompressedBatchInputStream != null) {
        uncompressedBatchInputStream.close();
      }

      // using byte array here rather than ByteBuf since the getMemqLogMessage will iterate through one message at a time,
      // which would be in the range of tens of KBs, unless the message itself has a extremely high compression ratio
      uncompressedBatchInputStream = CommonUtils.getUncompressedInputStream(header.getCompression(),
          new ByteArrayInputStream(batch));
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
    return messagesToRead > 0;
  }

  protected MemqMessageHeader readHeader() throws IOException {
    MemqMessageHeader header = new MemqMessageHeader(stream);
    if (header.getMessageLength() > objectSize) {
      logger.severe("BatchLength (" + header.getMessageLength() + ") is larger than objectSize ("
          + objectSize + ")");
      throw new IOException(
          "Invalid batchLength found:" + header.getMessageLength() + " vs:" + objectSize);
    } else if (header.getMessageLength() < 0) {
      logger.severe("BatchLength (" + header.getMessageLength() + ") is less than zero");
      throw new IOException("Invalid batchLength found:" + header.getMessageLength());
    }
    messagesToRead = header.getLogmessageCount();
    return header;
  }

  /**
   * This method allows skipping to last message of a Batch efficiently. <br>
   * Using this method allows clients to skip decompression of one message at the
   * time and simply scroll to the last message. <br>
   * <br>
   * This method skips over all Messages but the last one and then reads the last
   * message till (N-1)th message leaving the last message to be accessed via the
   * next() method.
   * 
   * @throws IOException
   */
  public void skipToLastLogMessage() throws IOException {
    currentMessageOffset = 0;
    if (stream == null) {
      throw new IOException("Stream is null can't skip to last");
    }
    while (stream.available() > 0) {
      readHeaderAndLoadBatch();
    }
    while (messagesToRead > 1) {
      getMemqLogMessage();
    }
  }

  /**
   * Reads the uncompressed input stream and returns a MemqLogMessage
   *
   * @return a MemqLogMessage
   * @throws IOException
   */
  protected MemqLogMessage<K, V> getMemqLogMessage() throws IOException {
    short logMessageInternalFieldsLength = uncompressedBatchInputStream.readShort();
    long writeTimestamp = 0;
    MessageId messageId = null;
    Map<String, byte[]> headers = null;
    if (logMessageInternalFieldsLength > 0) {
      writeTimestamp = uncompressedBatchInputStream.readLong();

      // messageId
      int messageIdLength = uncompressedBatchInputStream.read();
      if (messageIdLength > 0) {
        byte[] messageIdAry = new byte[messageIdLength];
        uncompressedBatchInputStream.readFully(messageIdAry);
        messageId = new MessageId(messageIdAry);
        messageIdHash = MemqUtils.calculateMessageIdHash(messageIdHash, messageId.toByteArray());
        auditedMessageCount++;
      }

      // headers
      short headerLength = uncompressedBatchInputStream.readShort();
      if (headerLength > 0) {
        headers = deserializeHeaders(headerLength, uncompressedBatchInputStream);
      }

      // ###################################################
      // do something with internal headers here in future
      // ###################################################
    }
    int keyLength = uncompressedBatchInputStream.readInt();
    byte[] keyBytes = null;
    if (keyLength > 0) {
      keyBytes = new byte[keyLength];
      uncompressedBatchInputStream.readFully(keyBytes);
    }
    int logMessageBytesToRead = uncompressedBatchInputStream.readInt();
    byte[] logMessageBytes = new byte[logMessageBytesToRead];
    uncompressedBatchInputStream.readFully(logMessageBytes);
    messagesToRead--;
    if (messagesToRead == 0) {
      tryAndSendAuditMessage();
    }
    metricRegistry.counter(MESSAGES_PROCESSED_COUNTER_KEY).inc();
    metricRegistry.meter(BYTES_PROCESSED_METER_KEY).mark(logMessageBytesToRead);
    MemqLogMessage<K, V> logMessage = new MemqLogMessage<>(messageId, headers,
        headerDeserializer.deserialize(keyBytes), valueDeserializer.deserialize(logMessageBytes), messagesToRead == 0);
    populateInternalFields(writeTimestamp, logMessage);
    currentMessageOffset++;
    return logMessage;
  }

  private void tryAndSendAuditMessage() throws IOException {
    // send audit message if the header and auditor are present
    // note that this logic basically represents that we send the message hash after
    // we have read all LogMessages from a given Message within a batch
    if (auditor != null && messageIdHash != null) {
      if (header != null) {
        auditor.auditMessage(cluster.getBytes(MemqUtils.CHARSET), topic.getBytes(MemqUtils.CHARSET),
            header.getProducerAddress(), header.getProducerEpoch(), header.getProducerRequestId(),
            messageIdHash, auditedMessageCount, false, clientId);
      } else {
        logger.warning("Header is null, we can't send audit");
      }
    }
  }

  private void populateInternalFields(long writeTimestamp, MemqLogMessage<K, V> logMessage) {
    logMessage.setWriteTimestamp(writeTimestamp);
    logMessage.setMessageOffsetInBatch(currentMessageOffset);
    try {
      logMessage.setNotificationPartitionId(notificationPartitionId);
      logMessage.setNotificationPartitionOffset(notificationPartitionOffset);
      logMessage.setNotificationReadTimestamp(notificationReadTimestamp);
    } catch (Exception e) {
    }
  }

  protected Map<String, byte[]> deserializeHeaders(short headerLength,
                                                   DataInputStream input) throws IOException {
    Map<String, byte[]> map = new HashMap<>();
    while (headerLength > 0) {
      short keyLength = input.readShort();
      byte[] key = new byte[keyLength];
      input.readFully(key);
      short valueLength = input.readShort();
      byte[] value = new byte[valueLength];
      input.readFully(value);
      map.put(new String(key), value);
      headerLength -= 4 + keyLength + valueLength;
    }
    return map;
  }

  @Override
  public void remove() {
    // do nothing
  }

  @Override
  public void forEachRemaining(Consumer<? super MemqLogMessage<K, V>> action) {
    while (hasNext()) {
      action.accept(next());
    }
  }

  @Override
  public void close() throws IOException {
    if (stream != null) {
      stream.close();
      stream = null;
    }
    if (uncompressedBatchInputStream != null) {
      uncompressedBatchInputStream.close();
      uncompressedBatchInputStream = null;
    }
  }
}

File: memq-python-client/src/memq/memqlogmessageiterator.py
'''
Created on Feb 2, 2021

@author: ambudsharma
'''

import collections
from memq.inputstream import DataInputStream
import gzip
from io import BytesIO
import logging
import zstandard


class MessageId:
    
    def __init__(self, array):
        self.array = array


class BatchHeader:
    '''
    classdocs
    '''

    def __init__(self, stream):
        '''
        Constructor
        '''
        stream.read_int();  # header length
        num_index_entries = stream.read_int();  # num index entries
        index = {}
        i = 0
        while i < num_index_entries:
            idx = stream.read_int()
            offset = stream.read_int()
            size = stream.read_int()
            index[idx] = IndexEntry(offset, size)
            i += 1
        self.messageIdx = collections.OrderedDict(index)
        logging.debug("Batch header index:" + str(self.messageIdx))


class IndexEntry:
    
    def __init__(self, offset, size):
        self.offset = offset
        self.size = size
        
    def __str__(self):
        return self.offset + " " + self.size


class MemqMessageHeader:
    
    def __init__(self, stream):
        self.headerLength = stream.read_short()
        self.verion = stream.read_short()
        self.addition_header_length = stream.read_short()
        if self.addition_header_length > 0:
            self.producer_address = stream.read(stream.read_byte())
            self.producer_epoch = stream.read_long()
            self.producer_request_id = stream.read_long()
        self.crc = stream.read_int()
        self.compression_type = stream.read_byte()
        self.log_message_count = stream.read_int()
        self.message_length = stream.read_int()
        
        attrs = vars(self)
        logging.info(', '.join("%s: %s" % item for item in attrs.items()))


class MemqLogMessage:
    '''
    classdocs
    '''
    INTERNAL_FIELD_TOPIC = "topic"
    INTERNAL_FIELD_WRITE_TIMESTAMP = "wts"
    INTERNAL_FIELD_NOTIFICATION_PARTITION_ID = "npi"
    INTERNAL_FIELD_NOTIFICATION_READ_TIMESTAMP = "nrts"
    INTERNAL_FIELD_NOTIFICATION_PARTITION_OFFSET = "npo"
    INTERNAL_FIELD_OBJECT_SIZE = "objectSize"
    INTERNAL_FIELD_MESSAGE_OFFSET = "mo"
    
    def __init__(self, message_id, headers, key, value):
        self.message_id = message_id
        self.headers = headers
        self.key = key
        self.value = value
        self.write_timestamp = -1
        self.message_offset_in_batch = -1
        self.notification_partition_id = -1
        self.notification_partition_offset = -1
        self.notification_read_timestamp = -1


class Commons:
    
    def checksum_matches(self, batch, crc):
        # run checksum match
        return True 
        
    def uncompress_stream(self, compression_type, batch_byte_stream):
        if compression_type == 0:
            # no compression
            return DataInputStream(batch_byte_stream)
        elif compression_type == 1:
            # gzip
            return DataInputStream(gzip.GzipFile(fileobj=batch_byte_stream, mode='rb'))
        elif compression_type == 2:
            # zstd
            zstd = zstandard.ZstdDecompressor();
            output_stream = BytesIO()
            zstd.copy_stream(batch_byte_stream, output_stream)
            output_stream.seek(0)
            return DataInputStream(output_stream)
        else:
            raise Exception("Unknown compression type")


class Deserializer:
    
    def deserialize(self, bytes):
        return bytes


class MemqLogMessageIterator(object):
    '''
    classdocs
    '''

    def __init__(self, cluster, stream, curr_notification_obj, header_deserializer, value_deserializer, skip_header_read=False):
        '''
        Constructor
        '''
        self.cluster = cluster
        self.stream = stream
        self.curr_notification_obj = curr_notification_obj
        try:
            self.topic = curr_notification_obj[MemqLogMessage.INTERNAL_FIELD_TOPIC]
            self.object_size = curr_notification_obj[MemqLogMessage.INTERNAL_FIELD_OBJECT_SIZE]
            self.notification_partition_id = curr_notification_obj[MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_PARTITION_ID]
            self.notification_partition_offset = curr_notification_obj[MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_PARTITION_OFFSET]
            self.notification_read_timestamp = curr_notification_obj[MemqLogMessage.INTERNAL_FIELD_NOTIFICATION_READ_TIMESTAMP]
        except Exception as e:
            # do nothing
            print(e)
        
        self.current_message_offset = 0
        self.header_deserializer = header_deserializer
        self.value_deserializer = value_deserializer
        
        # add metrics registry
        if skip_header_read == False:
            logging.debug("Parsing batch header")
            self.batch_header = BatchHeader(stream)
            self.num_batches = len(self.batch_header.messageIdx)
            logging.debug("Parsed batch header and found " + str(self.num_batches) + " batches")
        self.read_header_and_load_batch()
        
    def has_next(self) -> bool:
        # DONE
        try:
            if self.num_batches == 0:
                return self.messages_to_read > 0
        except Exception as e:
            raise e
        return True
    
    def next(self) -> MemqLogMessage:
        # DONE
        if self.messages_to_read > 0:
            try:
                return self.get_memq_log_message()
            except Exception as e:
                raise e
        else:
            try:
                if self.num_batches > 0:
                    if self.read_header_and_load_batch():
                        return self.next()
            except Exception as e:
                raise e
        raise Exception("No next")
        
    def read_header_and_load_batch(self) -> bool:
        '''
        Reads the next batch's header and returns whether there are bytes to read
        true if there are batch bytes to read, false otherwise
        '''
        try:
            self.message_id_hash = None
            self.header = self.read_header()
            batch_bytes = self.stream.read(self.header.message_length)
            if Commons.checksum_matches(self, batch_bytes, self.header.crc) == False:
                raise Exception("CRC checksum mismatch")
            
            self.uncompressed_batch_input_stream = Commons.uncompress_stream(self, compression_type=self.header.compression_type, batch_byte_stream=BytesIO(batch_bytes))
        except Exception as e:
            raise e
        return self.messages_to_read > 0
        
    def read_header(self) -> MemqMessageHeader:
        # DONE
        header = MemqMessageHeader(self.stream)
        logging.debug("Num batches " + str(self.num_batches))
        self.num_batches -= 1
        if header.message_length > self.object_size:
            raise Exception("Invalid batch length, length greater than object " + header.message_length)
        elif header.message_length < 0:
            raise Exception("Invalid batch length, length less than 0")
        else:
            self.messages_to_read = header.log_message_count
            return header
        
    def get_memq_log_message(self) -> MemqLogMessage:
        '''
        Reads the uncompressed input stream and returns a MemqLogMessage
        '''
        # DONE
        stream = self.uncompressed_batch_input_stream;
        log_message_internal_fields_length = stream.read_short()
        write_timestamp = 0
        message_id = None
        headers = None
        if log_message_internal_fields_length > 0:
            write_timestamp = stream.read_long()
            # message_id
            message_id_length = stream.read_byte()
            if  message_id_length > 0:
                array = stream.read(message_id_length)
                message_id = MessageId(array)
            
            # headers    
            header_length = stream.read_short()
            if header_length:
                headers = self.deserialize_headers(header_length, stream)
        
        # read key
        key_length = stream.read_int()
        key_bytes = None
        if key_length > 0:
            key_bytes = stream.read(key_length)
        
        # read value
        value_length = stream.read_int()
        value_bytes = stream.read(value_length)
        self.messages_to_read -= 1
        
        log_message = MemqLogMessage(message_id, headers, self.header_deserializer.deserialize(key_bytes), self.value_deserializer.deserialize(value_bytes))
        self.populate_internal_fields(write_timestamp, log_message)
        self.current_message_offset += 1
        return log_message
                
    def populate_internal_fields(self, write_timestamp, log_message):
        # DONE
        log_message.write_timestamp = write_timestamp
        log_message.message_offset_in_batch = self.current_message_offset
        log_message.notification_partition_id = self.notification_partition_id
        log_message.notification_partition_offset = self.notification_partition_offset
        log_message.notification_read_timestamp = self.notification_read_timestamp
    
    def deserialize_headers(self, header_length, stream) -> dict:
        # DONE
        header_map = {}
        while header_length > 0:
            key_length = stream.read_short()
            key = stream.read(key_length)
            value_length = stream.read_short()
            value = stream.read(value_length)
            header_map[key] = value
            header_length -= (4 + key_length + value_length)
        return header_map

